File: src/main/java/io/github/seggan/jyxal/compiler/wrappers/ContextualVariable.java
package io.github.seggan.jyxal.compiler.wrappers;

import org.objectweb.asm.Opcodes;

public class ContextualVariable implements AutoCloseable {

    final int index;
    private final JyxalMethod mv;

    private boolean isClosed = false;

    ContextualVariable(int index, JyxalMethod mv) {
        this.index = index;
        this.mv = mv;
    }

    @Override
    public void close() {
        checkClosed();
        mv.freeVar(this);
        isClosed = true;
    }

    public void load() {
        checkClosed();
        mv.visitVarInsn(Opcodes.ALOAD, index);
    }

    public void store() {
        checkClosed();
        mv.visitVarInsn(Opcodes.ASTORE, index);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ContextualVariable that)) return false;

        if (index != that.index) return false;
        return isClosed == that.isClosed && mv.equals(that.mv);
    }

    @Override
    public int hashCode() {
        int result = index;
        result = 31 * result + mv.hashCode();
        return result;
    }

    @Override
    public String toString() {
        return "ContextualVariable{index=%d, isClosed=%s}".formatted(index, isClosed);
    }

    private void checkClosed() {
        if (isClosed) {
            throw new IllegalStateException("Already closed");
        }
    }
}

File: src/main/kotlin/io/github/seggan/jyxal/compiler/wrappers/ContextualVariable.kt
package io.github.seggan.jyxal.compiler.wrappers

import org.objectweb.asm.Label
import org.objectweb.asm.Opcodes

data class ContextualVariable(internal val index: Int, private val mv: JyxalMethod) : AutoCloseable {

    internal val end = Label()

    private var isClosed = false

    override fun close() {
        checkClosed()
        mv.freeVar(this)
        isClosed = true
    }

    fun load(opcode: Int) {
        checkClosed()
        if (opcode != Opcodes.ALOAD && opcode != Opcodes.ILOAD && opcode != Opcodes.FLOAD && opcode != Opcodes.DLOAD && opcode != Opcodes.LLOAD) {
            throw IllegalArgumentException("Opcode must be a load opcode, but was $opcode")
        }
        mv.visitVarInsn(opcode, index)
    }

    fun load() = load(Opcodes.ALOAD)

    fun store(opcode: Int) {
        checkClosed()
        if (opcode != Opcodes.ASTORE && opcode != Opcodes.ISTORE && opcode != Opcodes.FSTORE && opcode != Opcodes.DSTORE && opcode != Opcodes.LSTORE) {
            throw IllegalArgumentException("Opcode must be a store opcode, but was $opcode")
        }
        mv.visitVarInsn(opcode, index)
    }

    fun store() = store(Opcodes.ASTORE)

    private fun checkClosed() {
        check(!isClosed) { "Already closed" }
    }
}
